lib/kitchen/transport/exec.rb

```diff
@@ -40,19 +40,105 @@ class Connection < Kitchen::Transport::Base::Connection
         def execute(command)
           return if command.nil?
 
-          run_command(command)
+          if host_os_windows?
+            run_command(run_from_file_command(command))
+            close
+          else
+            run_command(command)
+          end
+        end
+
+        def close
+          if host_os_windows?
+            FileUtils.remove(exec_script_file)
+          end
         end
 
         # "Upload" the files by copying them locally.
         #
         # @see Base#upload
         def upload(locals, remote)
-          FileUtils.mkdir_p(remote)
+          # evaluate $env:temp on Windows
+          real_remote = remote.to_s == "\$env:TEMP\\kitchen" ? kitchen_temp : remote
+          FileUtils.mkdir_p(real_remote)
           Array(locals).each do |local|
-            FileUtils.cp_r(local, remote)
+            FileUtils.cp_r(local, real_remote)
+          end
+        end
+
+        # (see Base#init_options)
+        def init_options(options)
+          super
+          @instance_name = @options.delete(:instance_name)
+          @kitchen_root = @options.delete(:kitchen_root)
+        end
+
+        private
+
+        # @return [String] display name for the associated instance
+        # @api private
+        attr_reader :instance_name
+
+        # @return [String] local path to the root of the project
+        # @api private
+        attr_reader :kitchen_root
+
+        # Takes a long command and saves it to a file and uploads it to
+        # the test instance. Windows has cli character limits.
+        #
+        # @param command [String] a long command to be saved and uploaded
+        # @return [String] a command that executes the uploaded script
+        # @api private
+        def run_from_file_command(command)
+          if logger.debug?
+            debug("Creating exec script for #{instance_name} (#{exec_script_file})")
+            debug("Executing #{exec_script_file}")
+          end
+          File.open(exec_script_file, "wb") { |file| file.write(command) }
+          %{powershell -file "#{exec_script_file}"}
```

Comment:
@mwrock I'm confused as all of the commands (such as installing Chef, etc) that are being given are powershell commands. Why would I put it in a `.bat` file? It seems to assume you need powershell.

Reply:
It's been a long time since I have looked at this code base so I might be missing something. It looks like `run_command` simply calls `Mixlib::ShellOut.new`. If so, sending that a powershell command would fail unless the command is prefixed with a call to powershell.exe with a `-command` arg. However in this case, you would end up with powershell.exe calling powershell.exe. That would indeed "work". So this is probably fine.